service/worker/scheduler/fx.go

```diff
@@ -0,0 +1,95 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package scheduler
+
+import (
+	"go.uber.org/fx"
+
+	"go.temporal.io/api/workflowservice/v1"
+	sdkworker "go.temporal.io/sdk/worker"
+	"go.temporal.io/sdk/workflow"
+	"go.temporal.io/server/api/historyservice/v1"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	workercommon "go.temporal.io/server/service/worker/common"
+)
+
+const (
+	WorkflowType  = "temporal-sys-scheduler-workflow"
+	TaskQueueName = "temporal-sys-scheduler-tq"
+)
+
+type (
+	workerComponent struct {
+		activityDeps
+	}
+
+	activityDeps struct {
+		fx.In
+		MetricsClient  metrics.Client
+		Logger         log.Logger
+		HistoryClient  historyservice.HistoryServiceClient
+		FrontendClient workflowservice.WorkflowServiceClient
+	}
+
+	fxResult struct {
+		fx.Out
+		Component workercommon.PerNSWorkerComponent `group:"perNamespaceWorkerComponent"`
+	}
+)
+
+var Module = fx.Options(
+	fx.Provide(NewResult),
+)
+
+func NewResult(params activityDeps) fxResult {
+	return fxResult{
+		Component: &workerComponent{
+			activityDeps: params,
+		},
+	}
+}
+
+func (s *workerComponent) DedicatedWorkerOptions() *workercommon.PerNSDedicatedWorkerOptions {
+	return &workercommon.PerNSDedicatedWorkerOptions{
+		TaskQueue: TaskQueueName,
+		// TODO: maybe make dynamic config
```

Comment:
we probably will have to do so, but in follow up PR.

Reply:
sort of. but in that case we'd be better off increasing the number of pollers in the worker, right? I'd rather just do it right

Reply:
anyway, I'll do this in the next pr that adds dynamic config

Reply:
There is difference. Having only one worker with increased poller limits your resource to only one worker pod. Having multiple workers make it randomly distribute them among available worker pods. It is possible that some of them would land on same pod, but if you increase the number the chance of having them distribute among multiple pods dramatically increased. Same problem exists in distributing history shards and task queue partitions. They are all randomly distributed among available pods. I don't think we should dig into the serviceResolver, we are changing the membership interface and likely remove ringpop in near future. Should not add new dependency to that before we finish the membership refactoring.

Reply:
sorry, I didn't explain this well. I meant we could do this:

ask for n nodes (by appending an index). if k of them fall on the same node, then instead of creating k sdk workers, create one sdk worker with 2*k pollers (assuming default is 2). it's actually easier in the code to do that than it is to create k sdk workers for the same namespace+component pair, and it seems better too.

but I'll still leave a TODO in there to do it right with ResolveN